### PR TITLE
SPARK-5960 allow aws credentials to be passed to KinesisUtils.createStream

### PR DESCRIPTION
**Pull request by mce**

### CODE REVIEW

File: extras/kinesis-asl/pom.xml [hunk 1]
```diff
@@ -48,6 +48,11 @@
       <scope>test</scope>
     </dependency>
     <dependency>
+      <groupId>com.fasterxml.jackson.core</groupId>
```
Comment (srowen) [inline extras/kinesis-asl/pom.xml]: (Why is this needed?)

File: extras/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/KinesisReceiver.scala [hunk 2]
```diff
@@ -70,6 +69,20 @@ private[kinesis] class KinesisReceiver(
     storageLevel: StorageLevel)
   extends Receiver[Array[Byte]](storageLevel) with Logging { receiver =>
 
+  def this(appName: String,
+           streamName: String,
+           endpointUrl: String,
+           checkpointInterval: Duration,
+           initialPositionInStream: InitialPositionInStream,
+           storageLevel: StorageLevel,
+           credentials: AWSCredentials) = {
```
Comment (srowen) [inline extras/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/KinesisReceiver.scala]: Since this is internal, can this just be implemented with an optional final parameter to the existing declared constructor?

File: extras/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/KinesisReceiver.scala [hunk 3]
```diff
@@ -119,7 +132,9 @@ private[kinesis] class KinesisReceiver(
    */
   override def onStart() {
     workerId = InetAddress.getLocalHost.getHostAddress() + ":" + UUID.randomUUID()
-    credentialsProvider = new DefaultAWSCredentialsProviderChain()
+    if(credentialsProvider == null) {
+      credentialsProvider = new DefaultAWSCredentialsProviderChain()
```
Comment (srowen) [inline extras/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/KinesisReceiver.scala]: Can this simply be the default value of the field rather than delay its init? or is this expensive?

File: extras/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/KinesisUtils.scala [hunk 4]
```diff
@@ -16,6 +16,7 @@
  */
 package org.apache.spark.streaming.kinesis
 
+import com.amazonaws.auth.AWSCredentials
```
Comment (srowen) [inline extras/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/KinesisUtils.scala]: Nit: space after import. 

I know we have all lived to regret including third-party classes as part of the API. Although it seems maybe essential here, I wonder if there is any reasonable way to not accept this as `AWSCredentials` but as access keys or something. I suppose that this could work but would assume a certain type of credentials are being used.

File: extras/kinesis-asl/pom.xml (hunk 1 above)
Comment (mce) [inline extras/kinesis-asl/pom.xml]: tests are failing on kinesis-acl, at least on my local. Error message is saying a method is missing from a class in jackson library. I don't know the exact reason but adding this dependency fixes the error

File: extras/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/KinesisUtils.scala (hunk 4 above)
Comment (mce) [inline extras/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/KinesisUtils.scala]: Updated code a bit after your comments. 

Created ```Credentials``` class to avoid third-party import in api.
Moved credentials parameter to existing constructor as an optional parameter.

File: extras/kinesis-asl/pom.xml (hunk 1 above)
Comment (srowen) [inline extras/kinesis-asl/pom.xml]: Hm, I doubt this is the right way to fix it, since you are declaring a compile-time dependency on something that you don't use. Surely it's because there is a missing dep on some Kinesis component at runtime scope? you also shouldn't manage versions here.

File: extras/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/Credentials.scala [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.spark.streaming.kinesis
+
+import com.amazonaws.auth.AWSCredentials
+
+case class Credentials(accessKeyId: String, secretKey: String) extends AWSCredentials {
```
Comment (srowen) [inline extras/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/Credentials.scala]: Hm, I wonder if that actually solves the problem since it still extends the AWS class. If the point is to pass access / secret key, that can be done by properties already right? so the point of this must be to enable other credential types? in which case it may be unavoidable to use the AWS class, and maybe fine given the nature of this module.

File: extras/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/KinesisUtils.scala [hunk 6]
```diff
@@ -90,8 +93,9 @@ object KinesisUtils {
       endpointUrl: String, 
       checkpointInterval: Duration,
       initialPositionInStream: InitialPositionInStream,
-      storageLevel: StorageLevel): JavaReceiverInputDStream[Array[Byte]] = {
+      storageLevel: StorageLevel,
```
Comment (srowen) [inline extras/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/KinesisUtils.scala]: This changes the Java API right? you have to pass credentials now.

**Comment by srowen** [comment]
ok to test

File: extras/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/KinesisUtils.scala (hunk 6 above)
Comment (mce) [inline extras/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/KinesisUtils.scala]: yes

**Comment by mce** [comment]
@srowen tests passed, fyi

File: extras/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/KinesisUtils.scala (hunk 6 above)
Comment (srowen) [inline extras/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/KinesisUtils.scala]: Yes, the idea is not to change the API. I suppose this is less important as a module, but still. I would have thought MiMa would flag this but I don't know that it checks this module.

I think this is really something for @cfregly et al to review

File: extras/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/KinesisReceiver.scala [hunk 7]
```diff
@@ -24,8 +24,7 @@ import org.apache.spark.storage.StorageLevel
 import org.apache.spark.streaming.Duration
 import org.apache.spark.streaming.receiver.Receiver
 
-import com.amazonaws.auth.AWSCredentialsProvider
-import com.amazonaws.auth.DefaultAWSCredentialsProviderChain
+import com.amazonaws.auth.{AWSCredentials, AWSCredentialsProvider, DefaultAWSCredentialsProviderChain}
 import com.amazonaws.services.kinesis.clientlibrary.interfaces.IRecordProcessor
```
Comment (cfregly) [inline extras/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/KinesisReceiver.scala]: collapse the IRecordProcessor and IRecordProcessorFactory into a single import, as well?

File: extras/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/KinesisReceiver.scala [hunk 8]
```diff
@@ -82,15 +82,19 @@ private[kinesis] class KinesisReceiver(
   var workerId: String = null
 
   /*
-   * This impl uses the DefaultAWSCredentialsProviderChain and searches for credentials 
-   *   in the following order of precedence:
+   * This impl uses the DefaultAWSCredentialsProviderChain unless it's provided by constructor
+   *  and searches for credentials in the following order of precedence:
```
Comment (cfregly) [inline extras/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/KinesisReceiver.scala]: might be good to reword this to be more explicit about the 2 different scenarios:
1) Some[AWSCredentials] are provided in which case we use those
2) None are provided and therefore DefaultAWSCredentialsProviderChain will be constructed which searches for credentials in the following order...

File: extras/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/KinesisReceiver.scala [hunk 9]
```diff
@@ -82,15 +82,19 @@ private[kinesis] class KinesisReceiver(
...
    * Environment Variables - AWS_ACCESS_KEY_ID and AWS_SECRET_KEY
    * Java System Properties - aws.accessKeyId and aws.secretKey
    * Credential profiles file at the default location (~/.aws/credentials) shared by all 
    *   AWS SDKs and the AWS CLI
    * Instance profile credentials delivered through the Amazon EC2 metadata service
    */
-  var credentialsProvider: AWSCredentialsProvider = null
+  var credentialsProvider: AWSCredentialsProvider = (credentials map { cr => new AWSCredentialsProvider {
```
Comment (cfregly) [inline extras/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/KinesisReceiver.scala]: i seem to remember having problems when constructing the AWSCredentialsProvider outside of the onStart() method due to this hierarchy (including DefaultAWSCredentialsProviderChain) not being Serializable.

i assume this works, but wondering what may have changed.

also, is there a simpler way to do this in Scala?  i feel like this could be simplified, but maybe i'm missing something.  seems like a lot of syntax for something so simple.

File: extras/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/KinesisUtils.scala (hunk 4 above)
Comment (cfregly) [inline extras/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/KinesisUtils.scala]: my initial attempt at adding this support introduced a BasicAWSCredentialsProvider as follows:

https://github.com/cfregly/spark/blob/kinesis-dbc/extras/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/BasicAWSCredentialsProvider.scala

highlights:
1) takes AccessKey and SecretKey as Strings
2) follows the same hierarchy as the rest of the AWSCredentialsProviders (including not being Serializable)

File: extras/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/KinesisUtils.scala [hunk 10]
```diff
@@ -59,9 +61,10 @@ object KinesisUtils {
       endpointUrl: String,
       checkpointInterval: Duration,
       initialPositionInStream: InitialPositionInStream,
-      storageLevel: StorageLevel): ReceiverInputDStream[Array[Byte]] = {
+      storageLevel: StorageLevel,
+      credentials: Option[AWSCredentials]): ReceiverInputDStream[Array[Byte]] = {
```
Comment (cfregly) [inline extras/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/KinesisUtils.scala]: any thought to using a default param where new DefaultAWSCredentialsProvider() is the default to reduce the API changes?

File: extras/kinesis-asl/src/test/scala/org/apache/spark/streaming/kinesis/KinesisReceiverSuite.scala [hunk 11]
```diff
@@ -86,7 +86,7 @@ class KinesisReceiverSuite extends TestSuiteBase with Matchers with BeforeAndAft
     // Tests the API, does not actually test data receiving
     val kinesisStream = KinesisUtils.createStream(ssc, "mySparkStream",
       "https://kinesis.us-west-2.amazonaws.com", Seconds(2),
-      InitialPositionInStream.LATEST, StorageLevel.MEMORY_AND_DISK_2);
+      InitialPositionInStream.LATEST, StorageLevel.MEMORY_AND_DISK_2, None)
```
Comment (cfregly) [inline extras/kinesis-asl/src/test/scala/org/apache/spark/streaming/kinesis/KinesisReceiverSuite.scala]: would be good to have tests around setting the credentials in the following 2 cases:
1) None (null)
2) Some[AWSCredentials]

File: extras/kinesis-asl/src/test/java/org/apache/spark/streaming/kinesis/JavaKinesisStreamSuite.java [hunk 12]
```diff
@@ -34,7 +34,7 @@ public void testKinesisStream() {
     // Tests the API, does not actually test data receiving
     JavaDStream<byte[]> kinesisStream = KinesisUtils.createStream(ssc, "mySparkStream",
         "https://kinesis.us-west-2.amazonaws.com", new Duration(2000), 
-        InitialPositionInStream.LATEST, StorageLevel.MEMORY_AND_DISK_2());
+        InitialPositionInStream.LATEST, StorageLevel.MEMORY_AND_DISK_2(), null);
```
Comment (cfregly) [inline extras/kinesis-asl/src/test/java/org/apache/spark/streaming/kinesis/JavaKinesisStreamSuite.java]: would be good to have tests around setting the credentials in the following 2 cases:
1) None (null)
2) Some[AWSCredentials]

File: extras/kinesis-asl/src/main/java/org/apache/spark/examples/streaming/JavaKinesisWordCountASL.java [hunk 13]
```diff
@@ -131,7 +131,7 @@ public static void main(String[] args) {
         for (int i = 0; i < numStreams; i++) {
           streamsList.add(
             KinesisUtils.createStream(jssc, streamName, endpointUrl, checkpointInterval, 
-            InitialPositionInStream.LATEST, StorageLevel.MEMORY_AND_DISK_2())
+            InitialPositionInStream.LATEST, StorageLevel.MEMORY_AND_DISK_2(), null)
```
Comment (cfregly) [inline extras/kinesis-asl/src/main/java/org/apache/spark/examples/streaming/JavaKinesisWordCountASL.java]: would be nice to have a true default here versus adding to the signature

File: extras/kinesis-asl/src/main/scala/org/apache/spark/examples/streaming/KinesisWordCountASL.scala [hunk 14]
```diff
@@ -113,7 +113,7 @@ private object KinesisWordCountASL extends Logging {
     /* Create the same number of Kinesis DStreams/Receivers as Kinesis stream's shards */
     val kinesisStreams = (0 until numStreams).map { i =>
       KinesisUtils.createStream(ssc, streamName, endpointUrl, kinesisCheckpointInterval,
-          InitialPositionInStream.LATEST, StorageLevel.MEMORY_AND_DISK_2)
+          InitialPositionInStream.LATEST, StorageLevel.MEMORY_AND_DISK_2, None)
```
Comment (cfregly) [inline extras/kinesis-asl/src/main/scala/org/apache/spark/examples/streaming/KinesisWordCountASL.scala]: would be nice to have a true default here versus adding to the signature

File: docs/streaming-kinesis-integration.md [hunk 15]
```diff
@@ -32,7 +32,7 @@ A Kinesis stream can be set up at one of the valid Kinesis endpoints with 1 or m
 		import com.amazonaws.services.kinesis.clientlibrary.lib.worker.InitialPositionInStream
 
 		val kinesisStream = KinesisUtils.createStream(
-        	streamingContext, [Kinesis stream name], [endpoint URL], [checkpoint interval], [initial position])
+        	streamingContext, [Kinesis stream name], [endpoint URL], [checkpoint interval], [initial position], [credentials])
```
Comment (cfregly) [inline docs/streaming-kinesis-integration.md]: might be good to include some more details on what this is, exactly.

this is a bit confusing if you're thinking of this in terms of ACCESS_KEY_ID and SECRET_KEY (2 separate strings)

also, be sure to include any imports that are relevant to this code sample - similar to the InitialPositionInStream reference right before it.

File: docs/streaming-kinesis-integration.md [hunk 16]
```diff
@@ -44,7 +44,7 @@ A Kinesis stream can be set up at one of the valid Kinesis endpoints with 1 or m
 		import com.amazonaws.services.kinesis.clientlibrary.lib.worker.InitialPositionInStream;
 
 		JavaReceiverInputDStream<byte[]> kinesisStream = KinesisUtils.createStream(
-        	streamingContext, [Kinesis stream name], [endpoint URL], [checkpoint interval], [initial position]);
+        	streamingContext, [Kinesis stream name], [endpoint URL], [checkpoint interval], [initial position], [credentials]);
```
Comment (cfregly) [inline docs/streaming-kinesis-integration.md]: same here :)

File: docs/streaming-kinesis-integration.md [hunk 17]
```diff
@@ -67,6 +67,8 @@ A Kinesis stream can be set up at one of the valid Kinesis endpoints with 1 or m
 
 	- `[initial position]`: Can be either `InitialPositionInStream.TRIM_HORIZON` or `InitialPositionInStream.LATEST` (see Kinesis Checkpointing section and Amazon Kinesis API documentation for more details).
 
+	- `[credentials]`: Optional AWS credentials.
```
Comment (cfregly) [inline docs/streaming-kinesis-integration.md]: it's a bit misleading to say this is optional since we still need to provide a value (None/null or Some)

File: extras/kinesis-asl/src/main/java/org/apache/spark/examples/streaming/JavaKinesisWordCountASL.java (hunk 13 above)
Comment (cfregly) [inline extras/kinesis-asl/src/main/java/org/apache/spark/examples/streaming/JavaKinesisWordCountASL.java]: also, might want to name this variable to be more explicit.  you can set it to None/null, but naming params goes a long way in these examples.

Comment (cfregly) [inline extras/kinesis-asl/src/main/java/org/apache/spark/examples/streaming/JavaKinesisWordCountASL.java]: another option is to require a non-null value to tighten up the interface - DefaultAWSCredentialsProvider() would be recommended, but people can pass in their own.

again, please make sure there are no Serializable issues if you go this route.  i remember having issues here.

File: extras/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/KinesisReceiver.scala (hunk 7 above)
Comment (mce) [inline extras/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/KinesisReceiver.scala]: done

File: extras/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/KinesisReceiver.scala (hunk 9 above)
Comment (mce) [inline extras/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/KinesisReceiver.scala]: i don't know if there's a simpler way but i'll think about it.

although the unit tests are passing now, i'll run example applications to check if the serialization issue you mentioned still exists

File: extras/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/KinesisUtils.scala (hunk 10 above)
Comment (mce) [inline extras/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/KinesisUtils.scala]: updated the parameter with default value ```None``` which means ```DefaultAWSCredentialsProvider``` will be used in the end.

However i couldn't to do the same for the Java api. Any suggestions?

File: extras/kinesis-asl/src/test/java/org/apache/spark/streaming/kinesis/JavaKinesisStreamSuite.java (hunk 12 above)
Comment (mce) [inline extras/kinesis-asl/src/test/java/org/apache/spark/streaming/kinesis/JavaKinesisStreamSuite.java]: added tests for both cases

File: extras/kinesis-asl/src/test/scala/org/apache/spark/streaming/kinesis/KinesisReceiverSuite.scala (hunk 11 above)
Comment (mce) [inline extras/kinesis-asl/src/test/scala/org/apache/spark/streaming/kinesis/KinesisReceiverSuite.scala]: added tests for both cases

File: docs/streaming-kinesis-integration.md (hunk 15 above)
Comment (mce) [inline docs/streaming-kinesis-integration.md]: added more explanation, please see line 72 https://github.com/vngrs/spark/commit/5cfd4377f432d821584a28b856948ebde9696339#diff-7f6eb703c33969d0218557112e1992b3R72

File: extras/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/KinesisReceiver.scala (hunk 9 above)
Comment (mce) [inline extras/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/KinesisReceiver.scala]: I just tested the new code by running the example applications and it seems the serialization problem still happening. I'll move the initialization into onStart method and run the tests again.

**Comment by mce** [comment]
@cfregly Moved initialization code into ```onStart``` method to fix serialization error. I tested by running example application and it runs without any problem.

**Comment by mce** [comment]
@cfregly any other comments?

File: extras/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/KinesisReceiver.scala [hunk 18]
```diff
@@ -68,7 +64,8 @@ private[kinesis] class KinesisReceiver(
     endpointUrl: String,
     checkpointInterval: Duration,
     initialPositionInStream: InitialPositionInStream,
-    storageLevel: StorageLevel)
+    storageLevel: StorageLevel,
+    credentials: Option[AWSCredentials])
```
Comment (tdas) [inline extras/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/KinesisReceiver.scala]: I am really not sure how this will work. AWSCredentials is not serializable, while the Receiver has to be serializable (as the receiver object is sent to the executors from the driver). Did you test whether this works?

**Comment by tdas** [comment]
I am taking a look at this PR a little late. But I am not sure how this work at all. AWSCredentials is not serializable.

**Comment by tdas** [comment]
@mce Are you able to update this? If you are not able to work on this, mind closing this PR?

**Comment by mce** [comment]
@tdas sorry for the late response, i'm closing the pr now as i don't have any time to work on this currently.
